### DISCUSSION TITLE
Limit a preloaded relation

### DISCUSSION THREAD

**Discussion by ad-on-is** [adonisjs/core#2320] [Help] [answered]
I want to query all `categories`, preload them with `products`, but limit the products to 10 items for each `category`

`return Category.query().preload('products', (query) => query.limit(10))`

Unfortunately, this limits the total amount of products to 10.

Trying it this way, tells me, that `Type Product[] is not assignable to type HasMany<typeof Product, LucidModel>`

```
const categories = await Category.all()
		categories.forEach(async (category) => {
			const products = await category
				.related('products')
				.query()
				.limit(10)

			category.products = products
		})
```

Could someone point me to the right direction?

**Comment by RomainLanz** [comment, marked as answer]
Hey @ad-on-is! 👋 

You can use the `groupLimit` method!
📚  https://preview.adonisjs.com/blog/august-2020-release#preloading-group-limit

```ts
Category
  .query()
  .preload('products', (query) => {
    query
      .groupOrderBy('products.id', 'desc')
      .groupLimit(10)
  })
```

Also, note that `Array#forEach` doesn't work with `async/await`.

**Reply by ad-on-is** [reply]
aaaaah, thank you, didn't know about that since it isn't mentioned in the guides 👍

**Reply by ad-on-is** [reply]
btw... I know about the forEach, and Promise.all() solution, but for simplicity sake, I omitted that.

**Reply by ad-on-is** [reply]
Oh, I just encountered, that when applying `groupLimit()` I cannot preload models related to products
```
Category
  .query()
  .preload('products', (query) => {
    query
      .groupOrderBy('products.id', 'desc')
      .groupLimit(10)
      .preload('user')
  })
```
